cpp/include/raft/linalg/mdarray.h

```diff
@@ -0,0 +1,394 @@
+ *   CUDA stream is honorred.
+ *
+ * - For container policy, we require an additional view method for it to be used in
+ *   mdspan.  The container policy reference implementation has multiple `access` methods
+ *   that accommodate needs for both mdarray and mdspan.  This is more difficult for us
+ *   since the policy might contain states that are unwanted inside a CUDA kernel.  Also,
+ *   on host we return a proxy to the actual value as `device_ref` so different access
+ *   methods will have different return type, which is less desirable.
+ */
+template <class ElementType, class Extents, class LayoutPolicy, class ContainerPolicy>
+class mdarray {
+ public:
+  using element_type = ElementType;
+  using extents_type = Extents;
+  using layout_type  = LayoutPolicy;
+  using mapping_type = typename layout_type::template mapping<extents_type>;
+
+  using index_type            = std::size_t;
+  using size_type             = std::size_t;
+  using difference_type       = std::ptrdiff_t;
+  using container_policy_type = ContainerPolicy;
+  using container_type        = typename container_policy_type::container_type;
+
+  static_assert(!std::is_const<ElementType>::value,
+                "Element type for container must not be const.");
+
+  using pointer         = typename container_policy_type::pointer;
+  using const_pointer   = typename container_policy_type::const_pointer;
+  using reference       = typename container_policy_type::reference;
+  using const_reference = typename container_policy_type::const_reference;
+
+  /**
+   * \brief the mdspan type returned by view method.
+   */
+  using view_type =
+    std::conditional_t<container_policy_type::is_host_type::value,
+                       host_mdspan<element_type,
+                                   extents_type,
+                                   layout_type,
+                                   typename container_policy_type::accessor_policy>,
+                       device_mdspan<element_type,
+                                     extents_type,
+                                     layout_type,
+                                     typename container_policy_type::accessor_policy>>;
+  using const_view_type =
+    std::conditional_t<container_policy_type::is_host_type::value,
+                       host_mdspan<element_type const,
+                                   extents_type,
+                                   layout_type,
+                                   typename container_policy_type::accessor_policy>,
+                       device_mdspan<element_type const,
+                                     extents_type,
+                                     layout_type,
+                                     typename container_policy_type::accessor_policy>>;
+
+ public:
+  constexpr mdarray() noexcept(std::is_nothrow_default_constructible<container_type>::value) =
+    default;
+  constexpr mdarray(mdarray const&) noexcept(
```

Comment:
It would be unfortunate if we would have to host our own implementation of `mdarray` just because of the container requires streams. Can you provide the current reasons behind the changes to the interface? As I understand it, you removed some constructors and the stream in the container policy is causing some issues but it would help to have more details

Reply:
@cjnolet I have made some comments in the document of this class for what changes I have made to the original mdarray lass, could you please take a look?

Reply:
I outlined the changes in the PR description.  Hope that helps.